ldap/servers/slapd/back-ldbm/archive.c

```diff
@@ -490,3 +490,212 @@ ldbm_back_ldbm2archive(Slapi_PBlock *pb)
+                            source, prerr, slapd_pr_strerror(prerr), return_value);
+            }
+            break;
+        }
+        bytes_to_write = return_value;
+        ptr = buffer;
+        for (i = 0; i < CPRETRY; i++) {
+            return_value = write(dest_fd, ptr, bytes_to_write);
+            if (return_value == bytes_to_write) {
+                break;
+            } else {
+                /* means error */
+                PRErrorCode prerr = PR_GetError();
+                if (task) {
+                    slapi_task_log_notice(task,
+                            "archive_copyfile - Failed to write (%s) error %d (%s) - real bytes %d, expected bytes: %d",
+                            dest_file, prerr, slapd_pr_strerror(prerr), return_value, bytes_to_write);
+                }
+                slapi_log_err(SLAPI_LOG_ERR, "archive_copyfile",
+                              "Failed to write (%s) error %d (%s) - real bytes %d, expected bytes: %d\n",
+                              dest_file, prerr, slapd_pr_strerror(prerr), return_value, bytes_to_write);
+
+                if (return_value > 0) {
+                    bytes_to_write -= return_value;
+                    ptr += return_value;
+                    slapi_log_err(SLAPI_LOG_NOTICE, "archive_copyfile",
+                                  "Retrying to write %d bytes\n", bytes_to_write);
+                    if (task) {
+                        slapi_task_log_notice(task,
+                                "archive_copyfile - Retrying to write %d bytes",
+                                bytes_to_write);
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+        if ((CPRETRY == i) || (return_value < 0)) {
+            return_value = -1;
+            break;
+        }
+    }
+
+error:
+    if (source_fd != -1) {
+        close(source_fd);
+    }
+    if (dest_fd != -1) {
+        close(dest_fd);
+    }
+    slapi_ch_free_string(&buffer);
+    slapi_ch_free_string(&dest_file);
+
+    return return_value;
+}
+
+static char *cert_files_600[] = {
+    "ca.crt", "Server-Cert.crt", "key4.db", "cert9.db", "pin.txt", "pwdfile.txt",
+    "noise.txt", NULL
+};
+static char *config_files_440[] = {"certmap.conf", "slapd-collations.conf", NULL};
+
+/* Archive nss files, 99user.ldif, and config files */
+int32_t
+ldbm_archive_config(char *bakdir, Slapi_Task *task)
+{
+    slapdFrontendConfig_t *config = getFrontendConfig();
+    char *schema_dir = config->schemadir;
+    char *cert_dir = config->certdir;
+    char *config_dir = config->configdir;
+    char *backup_config_dir = slapi_ch_smprintf("%s/config_files", bakdir);
+    char *dse_file = slapi_ch_smprintf("%s/dse.ldif", config_dir);
+    char *schema_file = slapi_ch_smprintf("%s/99user.ldif", schema_dir);
+    int32_t rc = -1;
+
+    /* Create config_files directory */
+    if (mkdir(backup_config_dir, 0700) != 0) {
```

Comment:
PR_mkdir ?

Reply:
This function was basically copied from the existing backup copy function.  Figured I didn't need to fully reinvent the wheel, but I can rewrite it...